<a href="https://colab.research.google.com/github/seanzhu815/424/blob/main/hw8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Q1

In [1]:
import pandas as pd

# Load the Excel file
excel_file = '/content/drive/MyDrive/411/interest_rate.xlsx'  # Replace with the actual path to your Excel file
df = pd.read_excel(excel_file)

# Calculate the average of a specific column
column_name = '1yr'  # Replace with the name of the column you want to calculate the average for
average = df[column_name].mean()

# Print the average
print("Average annual interest rate :", average)

Average annual interest rate : 3.8347826086956522


## GDP growth rate

In [3]:
import numpy as np

In [19]:
from tables.table import Column
import pandas as pd
import requests

# URL for World Bank API
url = "https://api.worldbank.org/v2/country/CHN/indicator/NY.GDP.MKTP.KD.ZG?format=json"

# Send a GET request to retrieve the data
response = requests.get(url)

# Extract the JSON data
data = response.json()

# Get the GDP growth rate data from the JSON
gdp_data = data[1]

# Convert the JSON data to a pandas DataFrame
df2 = pd.DataFrame(gdp_data)

# Print the DataFrame
df2=df2[['date', 'value']]
df2.iloc[0,1]=3.0
df2


date      value
0   2022   3.000000
1   2021   8.109793
2   2020   2.239702
3   2019   5.950501
4   2018   6.749774
5   2017   6.947201
6   2016   6.848762
7   2015   7.041329
8   2014   7.425764
9   2013   7.766150
10  2012   7.863736
11  2011   9.550832
12  2010  10.635871
13  2009   9.398726
14  2008   9.650679
15  2007  14.230861
16  2006  12.720956
17  2005  11.394592
18  2004  10.113621
19  2003  10.038030
20  2002   9.133631
21  2001   8.335733
22  2000   8.490093
23  1999   7.661652
24  1998   7.845952
25  1997   9.236780
26  1996   9.922557
27  1995  10.953954
28  1994  13.036807
29  1993  13.883729
30  1992  14.224530
31  1991   9.262786
32  1990   3.920251
33  1989   4.206334
34  1988  11.222595
35  1987  11.657428
36  1986   8.949962
37  1985  13.430678
38  1984  15.191540
39  1983  10.770203
40  1982   9.017114
41  1981   5.112761
42  1980   7.834145
43  1979   7.591325
44  1978  11.325978
45  1977   7.570000
46  1976  -1.570000
47  1975   8.720000
48  1974   2.310000
49  1973   7.760000

In [20]:
df2019_2022=df2.iloc[:4,]

df2019_2022

date     value
0  2022  3.000000
1  2021  8.109793
2  2020  2.239702
3  2019  5.950501

In [69]:
np.mean(df2019_2022.value)

4.824998797842845

##CSI300

In [5]:
csv_file = '/content/drive/MyDrive/411/Shanghai Shenzhen CSI 300 Historical Data.csv'
df3 = pd.read_csv(csv_file)
df3

Date     Price      Open      High       Low     Vol. Change %
0   06/01/2023  3,857.86  3,788.59  3,867.85  3,786.86  116.16K    1.56%
1   05/01/2023  3,798.54  4,031.09  4,092.85  3,781.40  126.39K   -5.72%
2   04/01/2023  4,029.09  4,059.62  4,169.67  3,925.44    3.14M   -0.54%
3   03/01/2023  4,050.93  4,068.73  4,137.25  3,935.02    3.49M   -0.46%
4   02/01/2023  4,069.46  4,170.13  4,207.41  4,026.45    2.33M   -2.10%
5   01/01/2023  4,156.86  3,864.84  4,268.15  3,831.25    1.80M    7.37%
6   12/01/2022  3,871.63  3,921.30  4,003.32  3,805.95    2.36M    0.48%
7   11/01/2022  3,853.04  3,519.82  3,877.68  3,517.62    2.92M    9.81%
8   10/01/2022  3,508.70  3,806.56  3,870.22  3,495.95    1.71M   -7.78%
9   09/01/2022  3,804.89  4,066.19  4,128.11  3,801.28    2.07M   -6.72%
10  08/01/2022  4,078.84  4,160.65  4,222.79  4,046.62    2.44M   -2.19%
11  07/01/2022  4,170.10  4,489.52  4,530.32  4,160.87    2.55M   -7.02%
12  06/01/2022  4,485.01  4,086.62  4,515.00  4,057.91    3.51M    9.62%
13  05/01/2022  4,091.52  3,995.35  4,096.20  3,808.57    2.36M    1.87%
14  04/01/2022  4,016.24  4,198.82  4,290.67  3,757.09    2.67M   -4.89%
15  03/01/2022  4,222.60  4,597.85  4,623.43  3,942.86    3.26M   -7.84%
16  02/01/2022  4,581.65  4,638.58  4,672.65  4,488.48    2.17M    0.39%
17  01/01/2022  4,563.77  4,957.98  4,961.45  4,559.83    2.72M   -7.62%
18  12/01/2021  4,940.37  4,831.89  5,143.84  4,823.93    3.49M    2.24%
19  11/01/2021  4,832.03  4,878.13  4,935.49  4,754.00    2.99M   -1.56%
20  10/01/2021  4,908.77  4,915.73  5,002.57  4,840.77    2.35M    0.87%
21  09/01/2021  4,866.38  4,804.69  5,042.80  4,763.30    4.61M    1.26%
22  08/01/2021  4,805.61  4,794.82  5,059.79  4,720.83    4.10M   -0.12%
23  07/01/2021  4,811.17  5,241.20  5,252.26  4,663.90    3.79M   -7.90%
24  06/01/2021  5,224.04  5,320.47  5,352.64  5,058.57    2.85M   -2.02%
25  05/01/2021  5,331.57  5,098.80  5,378.48  4,925.92    2.66M    4.06%
26  04/01/2021  5,123.49  5,059.94  5,181.15  4,900.30    2.75M    1.49%
27  03/01/2021  5,048.36  5,389.25  5,454.97  4,883.66    4.01M   -5.40%
28  02/01/2021  5,336.76  5,356.85  5,930.91  5,318.86    3.01M   -0.28%
29  01/01/2021  5,351.96  5,212.93  5,655.48  5,190.94    4.13M    2.70%
30  12/01/2020  5,211.29  4,963.13  5,215.62  4,849.35    3.29M    5.06%
31  11/01/2020  4,960.25  4,702.75  5,055.19  4,695.44    3.30M    5.64%
32  10/01/2020  4,695.33  4,663.61  4,849.60  4,639.30    1.83M    2.35%
33  09/01/2020  4,587.40  4,803.39  4,877.01  4,554.72    2.67M   -4.75%
34  08/01/2020  4,816.22  4,735.90  4,901.21  4,566.27    3.88M    2.58%
35  07/01/2020  4,695.05  4,172.64  4,878.08  4,163.50    5.95M   12.75%
36  06/01/2020  4,163.96  3,901.73  4,174.42  3,901.73    2.55M    7.68%
37  05/01/2020  3,867.02  3,867.26  3,998.10  3,800.18    1.79M   -1.16%
38  04/01/2020  3,912.58  3,682.30  3,921.70  3,652.32    2.24M    6.14%
39  03/01/2020  3,686.16  3,968.84  4,215.85  3,503.19    3.67M   -6.44%
40  02/01/2020  3,940.05  3,639.91  4,184.44  3,639.91    3.66M   -1.59%
41  01/01/2020  4,003.90  4,121.35  4,223.51  3,972.39    2.15M   -2.26%
42  12/01/2019  4,096.58  3,835.51  4,098.14  3,811.63    2.51M    7.00%
43  11/01/2019  3,828.67  3,883.81  4,030.64  3,804.83    1.85M   -1.49%
44  10/01/2019  3,886.75  3,818.59  3,983.81  3,805.07    1.47M    1.89%
45  09/01/2019  3,814.53  3,803.69  3,985.24  3,799.82    2.17M    0.39%
46  08/01/2019  3,799.59  3,819.32  3,848.24  3,575.86    2.10M   -0.93%
47  07/01/2019  3,835.36  3,899.33  3,942.43  3,745.85    2.13M    0.26%
48  06/01/2019  3,825.59  3,640.61  3,856.09  3,557.02    2.00M    5.39%
49  05/01/2019  3,629.79  3,775.08  3,791.61  3,556.26    2.26M   -7.24%

In [6]:
df3 = df3.replace(',', '', regex=True)
df3

Date    Price     Open     High      Low     Vol. Change %
0   06/01/2023  3857.86  3788.59  3867.85  3786.86  116.16K    1.56%
1   05/01/2023  3798.54  4031.09  4092.85  3781.40  126.39K   -5.72%
2   04/01/2023  4029.09  4059.62  4169.67  3925.44    3.14M   -0.54%
3   03/01/2023  4050.93  4068.73  4137.25  3935.02    3.49M   -0.46%
4   02/01/2023  4069.46  4170.13  4207.41  4026.45    2.33M   -2.10%
5   01/01/2023  4156.86  3864.84  4268.15  3831.25    1.80M    7.37%
6   12/01/2022  3871.63  3921.30  4003.32  3805.95    2.36M    0.48%
7   11/01/2022  3853.04  3519.82  3877.68  3517.62    2.92M    9.81%
8   10/01/2022  3508.70  3806.56  3870.22  3495.95    1.71M   -7.78%
9   09/01/2022  3804.89  4066.19  4128.11  3801.28    2.07M   -6.72%
10  08/01/2022  4078.84  4160.65  4222.79  4046.62    2.44M   -2.19%
11  07/01/2022  4170.10  4489.52  4530.32  4160.87    2.55M   -7.02%
12  06/01/2022  4485.01  4086.62  4515.00  4057.91    3.51M    9.62%
13  05/01/2022  4091.52  3995.35  4096.20  3808.57    2.36M    1.87%
14  04/01/2022  4016.24  4198.82  4290.67  3757.09    2.67M   -4.89%
15  03/01/2022  4222.60  4597.85  4623.43  3942.86    3.26M   -7.84%
16  02/01/2022  4581.65  4638.58  4672.65  4488.48    2.17M    0.39%
17  01/01/2022  4563.77  4957.98  4961.45  4559.83    2.72M   -7.62%
18  12/01/2021  4940.37  4831.89  5143.84  4823.93    3.49M    2.24%
19  11/01/2021  4832.03  4878.13  4935.49  4754.00    2.99M   -1.56%
20  10/01/2021  4908.77  4915.73  5002.57  4840.77    2.35M    0.87%
21  09/01/2021  4866.38  4804.69  5042.80  4763.30    4.61M    1.26%
22  08/01/2021  4805.61  4794.82  5059.79  4720.83    4.10M   -0.12%
23  07/01/2021  4811.17  5241.20  5252.26  4663.90    3.79M   -7.90%
24  06/01/2021  5224.04  5320.47  5352.64  5058.57    2.85M   -2.02%
25  05/01/2021  5331.57  5098.80  5378.48  4925.92    2.66M    4.06%
26  04/01/2021  5123.49  5059.94  5181.15  4900.30    2.75M    1.49%
27  03/01/2021  5048.36  5389.25  5454.97  4883.66    4.01M   -5.40%
28  02/01/2021  5336.76  5356.85  5930.91  5318.86    3.01M   -0.28%
29  01/01/2021  5351.96  5212.93  5655.48  5190.94    4.13M    2.70%
30  12/01/2020  5211.29  4963.13  5215.62  4849.35    3.29M    5.06%
31  11/01/2020  4960.25  4702.75  5055.19  4695.44    3.30M    5.64%
32  10/01/2020  4695.33  4663.61  4849.60  4639.30    1.83M    2.35%
33  09/01/2020  4587.40  4803.39  4877.01  4554.72    2.67M   -4.75%
34  08/01/2020  4816.22  4735.90  4901.21  4566.27    3.88M    2.58%
35  07/01/2020  4695.05  4172.64  4878.08  4163.50    5.95M   12.75%
36  06/01/2020  4163.96  3901.73  4174.42  3901.73    2.55M    7.68%
37  05/01/2020  3867.02  3867.26  3998.10  3800.18    1.79M   -1.16%
38  04/01/2020  3912.58  3682.30  3921.70  3652.32    2.24M    6.14%
39  03/01/2020  3686.16  3968.84  4215.85  3503.19    3.67M   -6.44%
40  02/01/2020  3940.05  3639.91  4184.44  3639.91    3.66M   -1.59%
41  01/01/2020  4003.90  4121.35  4223.51  3972.39    2.15M   -2.26%
42  12/01/2019  4096.58  3835.51  4098.14  3811.63    2.51M    7.00%
43  11/01/2019  3828.67  3883.81  4030.64  3804.83    1.85M   -1.49%
44  10/01/2019  3886.75  3818.59  3983.81  3805.07    1.47M    1.89%
45  09/01/2019  3814.53  3803.69  3985.24  3799.82    2.17M    0.39%
46  08/01/2019  3799.59  3819.32  3848.24  3575.86    2.10M   -0.93%
47  07/01/2019  3835.36  3899.33  3942.43  3745.85    2.13M    0.26%
48  06/01/2019  3825.59  3640.61  3856.09  3557.02    2.00M    5.39%
49  05/01/2019  3629.79  3775.08  3791.61  3556.26    2.26M   -7.24%

##Q2

rf = β + μc γ

Set β to 0 and γ to be 1

As for μc in china, it should be 4.82%(calculated before)

So the interest rate should be rf= 0 + 4.82%*γ=4.82% which is reasonable

##Q3

The average annual interest rate of china is 3.82% according to the calculation before. 

But according to CCAPM, the interest rate should be 4.82%.

It means that people consumed more than the china government predicted and means that the economics recession.

##Q4

In [62]:
stock_return_CSI300=(float(df3.iloc[0,2])/float(df3.iloc[-2,2])-1)*100/4
risk_premium=stock_return_CSI300-average
risk_premium

-2.8186067480569115

## Q5

In [8]:
df3.iloc[41,2]

'4121.35'

In [33]:
return_2019=(float(df3.iloc[41,2])/float(df3.iloc[-1,2])-1)
return_2019

0.0917252084724034

In [34]:
return_2020=(float(df3.iloc[29,2])/float(df3.iloc[-13,2])-1)

In [35]:
return_2021=(float(df3.iloc[17,2])/float(df3.iloc[-25,2])-1)

In [36]:
return_2022=(float(df3.iloc[5,2])/float(df3.iloc[-37,2])-1)

In [53]:
returns=np.array([return_2022,return_2021,return_2020,return_2019])

In [54]:
returns

array([-0.03266547, -0.02761826,  0.34796471,  0.09172521])

In [55]:
consumption_growth=np.array(df2019_2022['value'])

In [56]:
consumption_growth

array([3.        , 8.10979258, 2.23970186, 5.95050075])

In [57]:
gamma = 2

# Calculate the covariance between asset returns and consumption growth
covariance = np.cov(returns, consumption_growth)[0,1]

# Calculate the risk premium
risk_premium = gamma * covariance

print("Risk Premium:", risk_premium)

Risk Premium: -0.5516406965053231
